File: android/app/src/main/java/com/douxev/eggshell/ui/components/Charts.kt
package com.douxev.eggshell.ui.components

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.douxev.eggshell.punctuality.DeltaLabel
import com.douxev.eggshell.punctuality.DosePoint
import com.douxev.eggshell.punctuality.DoseTiming
import com.douxev.eggshell.punctuality.axisLabel
import com.douxev.eggshell.punctuality.punctualityAxis
import com.douxev.eggshell.punctuality.timingOf
import com.douxev.eggshell.ui.theme.EggColors

/**
 * Progress ring of the dose card. The 600 ms `cubic-bezier(.2,0,0,1)` curve is
 * the one the handoff specifies for the ring, not the generic M3 easing.
 */
@Composable
fun ProgressRing(
    value: Float,
    modifier: Modifier = Modifier,
    diameter: Dp = 64.dp,
    stroke: Dp = 6.dp,
    color: Color = MaterialTheme.colorScheme.primary,
    track: Color = color.copy(alpha = 0.20f),
    content: @Composable () -> Unit,
) {
    val animated by animateFloatAsState(
        targetValue = value.coerceIn(0f, 1f),
        animationSpec = tween(
            durationMillis = 600,
            easing = CubicBezierEasing(0.2f, 0f, 0f, 1f),
        ),
        label = "progress-ring",
    )
    Box(modifier = modifier.size(diameter), contentAlignment = Alignment.Center) {
        Canvas(modifier = Modifier.size(diameter)) {
            val w = stroke.toPx()
            val topLeft = Offset(w / 2, w / 2)
            val arcSize = Size(size.width - w, size.height - w)
            drawArc(
                color = track,
                startAngle = -90f,
                sweepAngle = 360f,
                useCenter = false,
                style = Stroke(width = w),
                topLeft = topLeft,
                size = arcSize,
            )
            drawArc(
                color = color,
                startAngle = -90f,
                sweepAngle = 360f * animated,
                useCenter = false,
                style = Stroke(width = w, cap = StrokeCap.Round),
                topLeft = topLeft,
                size = arcSize,
            )
        }
        content()
    }
}

/**
 * The punctuality chart (handoff §5.2) — new to this refonte, shared by Médics
 * and the doctor PDF.
 *
 * One dot per intake, X proportional to time. Y is the offset from the
 * prescribed time with `y = 0` at the top; the scale clamps to the largest
 * delay **of the period**. The legend is carried by the axis gradations, never
 * by a separate row under the plot.
 *
 * Strings arrive already localized so the component stays translation-agnostic:
 * [labelFor] renders an axis gradation and [missedLabel] the "oubliées · N" band.
 */
@Composable
fun PunctualityChart(
    points: List<DosePoint>,
    labelFor: (DeltaLabel) -> String,
    missedLabel: (Int) -> String,
    modifier: Modifier = Modifier,
    onTimeToleranceMin: Int = 15,
) {
    val axis = punctualityAxis(points)
    val scheme = MaterialTheme.colorScheme
    val gridColor = EggColors.chartGrid
    val measurer = rememberTextMeasurer()
    val labelStyle = MaterialTheme.typography.labelSmall

    val tickLabels = axis.ticks.map { labelFor(axisLabel(it)) }
    val missedText = missedLabel(axis.missedCount)
    val tickColors = listOf(scheme.tertiary, scheme.secondary, scheme.secondary)

    Canvas(modifier = modifier.fillMaxWidth()) {
        val gutter = 66.dp.toPx()
        val plotLeft = gutter + 8.dp.toPx()
        val plotRight = size.width
        val plotWidth = (plotRight - plotLeft).coerceAtLeast(1f)

        // Reserve a band for the missed doses under a dashed separator, and a
        // sliver of headroom above the zero line for the early intakes.
        val missedBandHeight = 18.dp.toPx()
        val separatorY = size.height - missedBandHeight
        val missedCenterY = size.height - missedBandHeight / 2
        val earlyHeadroom = if (axis.maxEarlyMin > 0) 10.dp.toPx() else 4.dp.toPx()
        val zeroY = earlyHeadroom
        val maxY = (separatorY - 6.dp.toPx()).coerceAtLeast(zeroY + 1f)

        fun yFor(deltaMin: Int): Float = when {
            deltaMin >= 0 ->
                zeroY + (deltaMin.toFloat() / axis.maxDelayMin) * (maxY - zeroY)
            axis.maxEarlyMin > 0 ->
                zeroY - (-deltaMin.toFloat() / axis.maxEarlyMin) * earlyHeadroom
            else -> zeroY
        }.coerceIn(0f, maxY)

        val dashed = PathEffect.dashPathEffect(floatArrayOf(4.dp.toPx(), 4.dp.toPx()))

        // Gradations. Zero is the dashed tertiary line; the other two are grid.
        axis.ticks.forEachIndexed { index, tick ->
            val y = yFor(tick)
            drawLine(
                color = if (index == 0) scheme.tertiary else gridColor,
                start = Offset(plotLeft, y),
                end = Offset(plotRight, y),
                strokeWidth = 1.dp.toPx(),
                pathEffect = if (index == 0) dashed else null,
            )
            val text = measurer.measure(
                tickLabels[index],
                style = labelStyle.merge(TextStyle(color = tickColors[index])),
            )
            drawText(
                textLayoutResult = text,
                topLeft = Offset(
                    x = gutter - text.size.width,
                    y = y - text.size.height / 2f,
                ),
            )
        }

        // Separator + the missed band's own axis label.
        drawLine(
            color = gridColor,
            start = Offset(plotLeft, separatorY),
            end = Offset(plotRight, separatorY),
            strokeWidth = 1.dp.toPx(),
            pathEffect = dashed,
        )
        val missedLayout = measurer.measure(
            missedText,
            style = labelStyle.merge(TextStyle(color = scheme.error)),
        )
        drawText(
            textLayoutResult = missedLayout,
            topLeft = Offset(
                x = gutter - missedLayout.size.width,
                y = missedCenterY - missedLayout.size.height / 2f,
            ),
        )

        if (points.isEmpty()) return@Canvas

        // X is proportional to time, never to the index.
        val firstMs = points.minOf { it.atMs }
        val lastMs = points.maxOf { it.atMs }
        val spanMs = (lastMs - firstMs).coerceAtLeast(1L)
        val radius = 2.6.dp.toPx()

        points.forEach { point ->
            val x = plotLeft + ((point.atMs - firstMs).toFloat() / spanMs) * plotWidth
            when (timingOf(point.deltaMin, onTimeToleranceMin)) {
                DoseTiming.Missed -> drawCircle(
                    color = scheme.error,
                    radius = radius,
                    center = Offset(x, missedCenterY),
                )
                DoseTiming.OnTime -> drawCircle(
                    color = scheme.tertiary,
                    radius = radius,
                    center = Offset(x, yFor(point.deltaMin ?: 0)),
                )
                DoseTiming.Late -> drawCircle(
                    color = scheme.secondary,
                    radius = radius,
                    center = Offset(x, yFor(point.deltaMin ?: 0)),
                )
            }
        }
    }
}

File: apple/Eggshell/Sources/Components/Charts.swift
import SwiftUI

// Charts of the refonte. One graphic vocabulary for the whole app (§5.1):
// primary = the main curve, tertiary = an on-time intake, secondary = a late
// one, error = a missed one, `chartGrid` = the grid. The legend is carried by
// the axis gradations, never by a separate row under the plot.

/// Progress ring of the dose card. The 600 ms `cubic-bezier(.2,0,0,1)` curve is
/// the one the handoff specifies for the ring, not a generic easing.
struct ProgressRingView<Center: View>: View {
    @Environment(\.palette) private var palette

    let progress: Double            // 0…1
    var diameter: CGFloat = 64
    var lineWidth: CGFloat = 6
    var tint: Color? = nil
    var track: Color? = nil
    @ViewBuilder var center: () -> Center

    @State private var shown: Double = 0

    var body: some View {
        ZStack {
            Circle()
                .stroke(track ?? palette.surfaceContainerHighest, lineWidth: lineWidth)
            Circle()
                .trim(from: 0, to: shown)
                .stroke(
                    tint ?? palette.primary,
                    style: StrokeStyle(lineWidth: lineWidth, lineCap: .round))
                .rotationEffect(.degrees(-90))
            center()
        }
        .frame(width: diameter, height: diameter)
        .onAppear { withAnimation(Self.curve) { shown = clamped } }
        .onChange(of: progress) { _, _ in
            withAnimation(Self.curve) { shown = clamped }
        }
    }

    private var clamped: Double { min(1, max(0, progress)) }
    private static var curve: Animation { .timingCurve(0.2, 0, 0, 1, duration: 0.6) }
}

/// The punctuality chart (§5.2) — new to this refonte, shared by Médics and the
/// doctor report.
///
/// One dot per intake, X proportional to time (never the index). Y is the
/// offset from the prescribed time with `y = 0` at the top; the scale clamps to
/// the largest delay **of the period**. Under a dashed separator sits the band
/// of the missed doses, with its own axis label.
struct PunctualityChartView: View {
    @Environment(\.palette) private var palette

    let points: [DosePoint]
    var height: CGFloat = 96
    var onTimeToleranceMin: Int = Punctuality.onTimeToleranceMin

    /// Width of the axis-label column, left of the plot.
    private let gutter: CGFloat = 62
    /// Height of the band the missed doses live in.
    private let missedBand: CGFloat = 18

    var body: some View {
        let axis = Punctuality.axis(points)
        Canvas { context, size in
            draw(axis: axis, in: context, size: size)
        }
        .frame(height: height)
        .frame(maxWidth: .infinity)
        .accessibilityElement()
        .accessibilityLabel(accessibilityText(axis))
    }

    private func draw(axis: PunctualityAxis, in context: GraphicsContext, size: CGSize) {
        let plotLeft = gutter + 8
        let plotRight = max(plotLeft + 1, size.width)
        let plotWidth = plotRight - plotLeft

        let separatorY = size.height - missedBand
        let missedCenterY = size.height - missedBand / 2
        // A sliver of headroom above the zero line so an early dose sits just
        // above it instead of escaping the plot.
        let earlyHeadroom: CGFloat = axis.maxEarlyMin > 0 ? 10 : 4
        let zeroY = earlyHeadroom
        let maxY = max(separatorY - 6, zeroY + 1)

        func yFor(_ deltaMin: Int) -> CGFloat {
            let raw: CGFloat
            if deltaMin >= 0 {
                raw = zeroY + CGFloat(deltaMin) / CGFloat(axis.maxDelayMin) * (maxY - zeroY)
            } else if axis.maxEarlyMin > 0 {
                raw = zeroY - CGFloat(-deltaMin) / CGFloat(axis.maxEarlyMin) * earlyHeadroom
            } else {
                raw = zeroY
            }
            return min(maxY, max(0, raw))
        }

        let dashed = StrokeStyle(lineWidth: 1, dash: [4, 4])
        let tickColors = [palette.tertiary, palette.secondary, palette.secondary]

        // Gradations. Zero is the dashed tertiary line, the other two are grid.
        for (index, tick) in axis.ticks.enumerated() {
            let y = yFor(tick)
            var line = Path()
            line.move(to: CGPoint(x: plotLeft, y: y))
            line.addLine(to: CGPoint(x: plotRight, y: y))
            context.stroke(
                line,
                with: .color(index == 0 ? palette.tertiary : palette.chartGrid),
                style: index == 0 ? dashed : StrokeStyle(lineWidth: 1))
            context.draw(
                Text(Punctuality.text(Punctuality.axisLabel(tick)))
                    .font(EggFont.micro)
                    .foregroundStyle(tickColors[min(index, tickColors.count - 1)]),
                at: CGPoint(x: gutter - 5, y: y),
                anchor: .trailing)
        }

        // Separator + the missed band's own axis label.
        var separator = Path()
        separator.move(to: CGPoint(x: plotLeft, y: separatorY))
        separator.addLine(to: CGPoint(x: plotRight, y: separatorY))
        context.stroke(separator, with: .color(palette.chartGrid), style: dashed)
        context.draw(
            Text(Punctuality.missedAxisText(axis.missedCount))
                .font(EggFont.micro)
                .foregroundStyle(palette.error),
            at: CGPoint(x: gutter - 5, y: missedCenterY),
            anchor: .trailing)

        guard !points.isEmpty else { return }

        // X is proportional to time, never to the index (§5.1).
        let firstMs = points.map(\.atMs).min() ?? 0
        let lastMs = points.map(\.atMs).max() ?? firstMs
        let spanMs = max(1, lastMs - firstMs)
        let radius: CGFloat = 3.1

        for point in points {
            let x = plotLeft + CGFloat(Double(point.atMs - firstMs) / Double(spanMs)) * plotWidth
            let timing = Punctuality.timing(point.deltaMin, onTimeToleranceMin: onTimeToleranceMin)
            let y: CGFloat
            let color: Color
            switch timing {
            case .missed:
                y = missedCenterY
                color = palette.error
            case .onTime:
                y = yFor(point.deltaMin ?? 0)
                color = palette.tertiary
            case .late:
                y = yFor(point.deltaMin ?? 0)
                color = palette.secondary
            }
            let dot = Path(ellipseIn: CGRect(
                x: x - radius, y: y - radius, width: radius * 2, height: radius * 2))
            context.fill(dot, with: .color(color))
        }
    }

    /// The chart's text alternative (§10): a canvas has no readable content of
    /// its own, so the whole thing is announced as one sentence.
    private func accessibilityText(_ axis: PunctualityAxis) -> String {
        let onTime = points.filter {
            Punctuality.timing($0.deltaMin, onTimeToleranceMin: onTimeToleranceMin) == .onTime
        }.count
        let late = points.filter {
            Punctuality.timing($0.deltaMin, onTimeToleranceMin: onTimeToleranceMin) == .late
        }.count
        return "Écart à l'heure prévue : \(onTime) à l'heure, \(late) en retard, "
            + "\(axis.missedCount) oubliées."
    }
}
